File: web/src/types/aiProviderUsage.ts
export interface AIProviderUsageRow {
  provider_id: string | null
  provider_version: number | null
  provider_name: string
  model: string
  total_requests: number
  successful_requests: number
  failed_requests: number
  success_rate_pct: number
  prompt_tokens: number
  completion_tokens: number
  total_tokens: number
  unknown_token_requests: number
  average_latency_ms: number | null
  p95_latency_ms: number | null
  not_sent_requests: number
  ambiguous_requests: number
  deadline_failures: number
  failure_categories: Record<string, number>
  last_request_at: string
}

export interface AIProviderUsageResponse {
  items: AIProviderUsageRow[]
  total: number
  days: number
  limit: number
  offset: number
}

File: backend/app/schemas/ai_provider_usage.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel


class AIProviderUsageRow(BaseModel):
    provider_id: uuid.UUID | None
    provider_version: int | None
    provider_name: str
    model: str
    total_requests: int
    successful_requests: int
    failed_requests: int
    success_rate_pct: float
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int
    unknown_token_requests: int
    average_latency_ms: float | None
    p95_latency_ms: float | None
    not_sent_requests: int
    ambiguous_requests: int
    deadline_failures: int
    failure_categories: dict[str, int]
    last_request_at: datetime


class AIProviderUsageResponse(BaseModel):
    items: list[AIProviderUsageRow]
    total: int
    days: int
    limit: int
    offset: int
